pkg/collector/pipeline_test.go

```diff
@@ -121,3 +123,89 @@ func TestDumpReports(t *testing.T) {
 		})
 	}
 }
+
+// Custom annotations
+
+var clientCountries = map[string]string{
+	"192.0.2.1": "US",
+	"192.0.2.2": "CA",
+}
+
+var serverZones = map[string]string{
+	"203.0.113.75": "us-east1-a",
+	"203.0.113.76": "us-west1-b",
+}
+
+type geoAnnotator struct{}
+
+func (g geoAnnotator) ProcessReports(batch *ReportBatch) {
+	batch.Annotation = clientCountries[batch.ClientIP]
+	for i, _ := range batch.Reports {
+		batch.Reports[i].Annotation = serverZones[batch.Reports[i].ServerIP]
+	}
+}
+
+type stashReports struct {
+	dest *ReportBatch
+}
+
+func (s stashReports) ProcessReports(batch *ReportBatch) {
+	*s.dest = *batch
+}
+
+func TestCustomAnnotation(t *testing.T) {
+	var got ReportBatch
+	pipeline := NewTestPipeline(newSimulatedClock())
+	pipeline.AddProcessor(&geoAnnotator{})
+	pipeline.AddProcessor(&stashReports{&got})
+	json := testdata(t, "annotation-nel-reports.json")
+
+	request := httptest.NewRequest("POST", "https://example.com/upload/", bytes.NewReader(json))
+	request.Header.Add("Content-Type", "application/report")
+	var response httptest.ResponseRecorder
+	pipeline.ServeHTTP(&response, request)
+
+	if response.Code != http.StatusNoContent {
+		t.Errorf("ServeHTTP(%s): got %d, wanted %d", compactJSON(json), response.Code, http.StatusNoContent)
+		return
+	}
+
+	want := ReportBatch{
```

Comment:
It feels a little weird to have the input in a file and the expected output hard coded here. Would it be appropriate to marshal the output back into a file so you can have a golden output? Or would that be too much extra functionality for the test?

Reply:
I like that idea!  Done.